osbuild/util/pathfd.py

```diff
@@ -0,0 +1,716 @@
+
+        return self._fd
+
+    def is_open(self):
+        """Check whether the Path Descriptor is Open
+
+        This simply returns a boolean that tells whether the descriptor is
+        still open. Note that once a descriptor is closed, it cannot be
+        re-opened. You must create a new descriptor to do that.
+
+        This can be safely called on closed descriptors.
+
+        Returns
+        -------
+        Boolean
+            Whether the descriptor is open. This is true for all descriptors
+            until they are closed via `close()`, or if they were created as a
+            closed descriptor.
+        """
+
+        return self._fd is not None
+
+    def clone(self):
+        """Create a Clone
+
+        This clones the given path-descriptor and returns it. The cloned
+        instance will point to the exact same resource as the original, without
+        re-resolving the path.
+
+        The original and the cloned instance do not share any underlying
+        resources. They are fully independent.
+
+        This can be safely called on closed descriptors, in which case the
+        cloned instance will also be closed.
+
+        Returns
+        -------
+        PathFd
+            A new path-desciptor for the same path is returned.
+
+        Raises
+        ------
+        OSError
+            Any operating system errors involved in duplicating
+            file-descriptors can be raised by this function. See `os.dup()` for
+            a discussion of possible errors.
+        """
+
+        if self.is_open():
+            return PathFd(_os.dup(int(self)))
+        else:
+            return PathFd()
+
+    def stat(self):
+        """Query File System Stat-Information
+
+        This performs the `fstat(2)` syscall on the file-system resource this
+        path-descriptor points to.
+
+        This is safe against TOCTOU races. Repeated calls are guaranteed to
+        return the same information, unless the information itself got changed
+        in the meantime.
+
+        This function does not cache any results. Every call results in a new
+        syscall invocation.
+
+        Note that this does not perform any path resolution. It accesses
+        exactly the resource the descriptor points to. If it points to a
+        symlink, it will return information for that symlink, rather than
+        information for the object the symlink points to. You must resolve
+        symlinks when creating a path-descriptor, in case you want symlinks
+        resolved.
+
+        Returns
+        -------
+        stat_result
+            The file-system information is reported as a return value of the
+            `stat_result` type from the python standard library.
+
+        Raises
+        ------
+        OSError
+            See the python standard library for possible errors when querying
+            stat-information (e.g., `os.stat()`).
+        """
+
+        assert self.is_open()
```

Comment:
I would expect a `ValueError("i/o operation on a closed path")`, like when you do i/o on a [closed file](https://github.com/python/cpython/blob/942f7a2dea2e95a0fa848329565c0d0288d92e47/Lib/_pyio.py#L486)